packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js

```diff
@@ -162,30 +127,74 @@ export const render = async (req, res, next) => {
     const component = await route.component.getComponent()
 
     // Step 3 - Init the app state
-    const {appState, error: appStateError} = await initAppState({
-        App: WrappedApp,
-        component,
-        match,
-        route,
-        req,
-        res,
-        location
-    })
+    let appState = {}
+    let routerContext = {}
+    let error
+    let routeError
+    let appStateError
+    let html
+
+    // Step 3. Bail if there is a 404.
+    if (component === Throw404) {
+        routeError = new errors.HTTPNotFound('Not found')
+    }
+
+    if (!routeError) {
+        const AppJSX = getAppJSX(req, res, error, {
+            App: WrappedApp,
+            appState,
+            deviceType,
+            location,
+            routerContext,
+            routes
+        })
+
+        try {
+            // Get all the data fetching promises for the various API's attached to the App component.
+            // This is the react query, and getProps enhancements.
+            const allPromises = WrappedApp.getDataPromises({
+                App,
+                AppJSX,
+                location,
+                req,
+                res,
+                deviceType,
+                route,
+                routes,
+                match
+            })
+            const data = await Promise.all(allPromises)
+
+            appState = data.reduce(
+                (acc, appState = {}) => ({
+                    ...acc,
+                    ...appState
+                }),
+                {}
+            )
+        } catch (e) {
+            appStateError = logAndFormatError(e || new Error())
```

Comment:
Querying Commerce API for a non-existent category, for example, would throw an error, which would get caught in this catch statement. The end result is this internal server error:

![Google Chrome 2022-08-30 at 16 50 48](https://user-images.githubusercontent.com/847300/187563315-d5da5172-973b-44a5-8259-8c6b88101c9b.png)



But what if the developers would want to catch and handle it in their app? Currently they can't because it's caught in the react rendering.

Reply:
One scenario I'm investigating is for the useExpress hook: how to detect a non-existent category and spit out a 404. I would want to be able to do the following in the user land:

```jsx
    const {categoryId} = useParams()
    const {data, isLoading, error} = useCategory({id: categoryId, levels: 0})
    const {req, res: pageRes} = useExpress()

    if (pageRes && error?.response?.status === 404) {
        pageRes.status(404).send(data?.detail)
    }
```